File: src/leetcode/P028_ImplementStrStr.java
package leetcode;

/**
 * 	Implement strStr().
 *  
 *  Returns the index of the first occurrence of needle in haystack,
 *  or -1 if needle is not part of haystack.
 */

public class P028_ImplementStrStr {
    //AC 13ms 34.88%
	static class Solution {
		public int strStr(String s, String p) {
		    int sn = s == null ? 0 : s.length();
		    int pn = p == null ? 0 : p.length();
		    //pn == 0, p match at index 0
		    if (pn == 0) return 0;
		    //sn == 0 and pn > 0 can not match
		    if (sn == 0) return -1;
		    int si = 0, pi = 0, next[] = getNext(p, pn);
		    while (si < sn) {
		        if (s.charAt(si) == p.charAt(pi)) {
		            si ++;
		            pi ++;
		            if (pi == pn) break;
		        } else if (next[pi] == -1) {
		            si ++;
		        } else pi = next[pi];
		    }
		    return pi == pn ? si - pn : - 1;
		}
		public int[] getNext(String p, int pn) {
		    //return short p
		    if (pn < 2) return new int[] {-1};
		    int bi = 0, fi = 2, next[] = new int[pn];
		    next[0] = -1;
		    next[1] = 0;
		    while (fi < pn) {
		        if (p.charAt(fi - 1) == p.charAt(bi)) {
		            next[fi ++] = ++ bi;
		        } else if (bi <= 0) {
		            next[fi ++] = 0;
		        } else bi = next[bi];
		    }
		    return next;
		}
	}
}

File: python/P028_ImplementStrStr.py
#coding=utf-8

'''
    url: leetcode.com/problems/implement-strstr/
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年4月1日
    @details:    Solution: 62ms 32.59%
'''

class Solution(object):
    def getNext(self, p, pn):
        #return short p
        if pn < 2: return [-1]
        bi, fi, ne = 0, 2, [0] * pn
        ne[0], ne[1] = -1, 0
        while fi < pn:
            if p[fi - 1] == p[bi]:
                bi += 1
                ne[fi] = bi
                fi += 1
            elif bi <= 0:
                ne[fi] = 0
                fi += 1
            else:
                bi = ne[bi]
        return ne
        
    def strStr(self, s, p):
        """
        :type s: str origin string
        :type p: str pattern string
        :rtype: int
        """
        sn = 0 if s == None else len(s)
        pn = 0 if p == None else len(p)
        #pn == 0 match s at index 0
        if pn == 0: return 0
        #sn == 0 and pn > 0 can not match
        if sn == 0: return -1
        si, pi, ne = 0, 0, self.getNext(p, pn)
        while si < sn:
            if s[si] == p[pi]:
                si += 1
                pi += 1
                if pi == pn: break
            elif ne[pi] == -1:
                si += 1
            else:
                pi = ne[pi]
        return si - pn if pi == pn else -1
    
if __name__ == "__main__":
    s, p ="mississippi", "issi"
    print(Solution().strStr(s, p))
    
    